### PR TITLE
NFC: codegen: eliminate builder global

### PR DESCRIPTION
**Pull request by vtjnash**
No functional change intended here (other than making the ctx argument usage and naming consistent). This just does the tedious work of passing jl_codectx_t to (almost) every function where it is needed. But since this ends up touching almost everything in codegen, I would like to merge it fairly quickly (before starting on removing the next bits of global state).

### CODE REVIEW

File: src/ccall.cpp [hunk 1]
```diff
@@ -208,23 +216,20 @@ static DenseMap<AttributeSet,
...
+        GlobalVariable *libptrgv, GlobalVariable *llvmgv,
+        void *symaddr, bool runtime_lib)
 {
     PointerType *funcptype = PointerType::get(functype, 0);
-    libptrgv = prepare_global(libptrgv, M);
-    llvmgv = prepare_global(llvmgv, M);
-    BasicBlock *old = builder.GetInsertBlock();
-    DebugLoc olddl = builder.getCurrentDebugLocation();
-    DebugLoc noDbg;
-    builder.SetCurrentDebugLocation(noDbg);
```
Comment (tkelman) [inline src/ccall.cpp]: why is this not necessary any more?

File: src/ccall.cpp [hunk 2]
```diff
@@ -1879,15 +1892,15 @@ jl_cgval_t function_sig_t::emit_a_ccall(
         jl_cgval_t *argv,
         SmallVector<Value*, 16> &gc_uses,
         bool static_rt,
-        jl_codectx_t *ctx)
+        jl_codectx_t &ctx)
```
Comment (tkelman) [inline src/ccall.cpp]: why is the argument order changed for every function except this one?

File: src/ccall.cpp [hunk 3]
```diff
@@ -1590,22 +1599,24 @@ static jl_cgval_t emit_ccall(jl_value_t **args, size_t nargs, jl_codectx_t *ctx)
...
-            ary = emit_unbox(largty, emit_expr(argi, ctx), tti);
+            ary = emit_unbox(ctx, largty, emit_expr(ctx, argi), tti);
         }
         JL_GC_POP();
         if (!retboxed) {
             return mark_or_box_ccall_result(
-                      emit_bitcast(emit_pointer_from_objref(
-                                          emit_bitcast(ary, T_prjlvalue)), lrt),
-                   retboxed, rt, unionall, static_rt, ctx);
+                    ctx,
```
Comment (tkelman) [inline src/ccall.cpp]: this indentation seems pretty arbitrary for a call site

File: src/ccall.cpp [hunk 4]
```diff
@@ -1653,99 +1664,101 @@ static jl_cgval_t emit_ccall(jl_value_t **args, size_t nargs, jl_codectx_t *ctx)
...
-        error_unless(builder.CreateICmpNE(defer_sig,
-                                          ConstantInt::get(T_sigatomic, 0)),
-                     "sigatomic_end called in non-sigatomic region", ctx);
-        defer_sig = builder.CreateSub(defer_sig,
-                                      ConstantInt::get(T_sigatomic, 1));
-        builder.CreateStore(defer_sig, pdefer_sig);
+        Value *defer_sig = ctx.builder.CreateLoad(pdefer_sig);
+        emit_signal_fence(ctx);
+        error_unless(ctx,
+                ctx.builder.CreateICmpNE(defer_sig, ConstantInt::get(T_sigatomic, 0)),
```
Comment (tkelman) [inline src/ccall.cpp]: why 8-space at a call site?

File: src/cgutils.cpp [hunk 5]
```diff
@@ -741,194 +739,185 @@ static jl_cgval_t emit_typeof(const jl_cgval_t &p, jl_codectx_t *ctx)
...
 
-static Value *emit_datatype_types(Value *dt)
+static Value *emit_datatype_types(jl_codectx_t &ctx, Value *dt)
 {
-    return tbaa_decorate(tbaa_const, builder.
-        CreateLoad(emit_bitcast(builder.
-                                CreateGEP(emit_bitcast(decay_derived(dt), T_pint8),
-                                          ConstantInt::get(T_size, offsetof(jl_datatype_t, types))),
-                                T_ppjlvalue)));
+    Value *Ptr = emit_bitcast(ctx, decay_derived(dt), T_ppjlvalue);
```
Comment (tkelman) [inline src/cgutils.cpp]: used to be T_pint8, why this change?

File: src/cgutils.cpp [hunk 6]
```diff
@@ -741,194 +739,185 @@ static jl_cgval_t emit_typeof(const jl_cgval_t &p, jl_codectx_t *ctx)
...
 {
-    Value *abstract = tbaa_decorate(tbaa_const, builder.
-        CreateLoad(builder.CreateGEP(emit_bitcast(decay_derived(dt), T_pint8),
-                                     ConstantInt::get(T_size, offsetof(jl_datatype_t, abstract)))));
-    return builder.CreateTrunc(abstract, T_int1);
+    Value *Ptr = emit_bitcast(ctx, decay_derived(dt), T_pint8);
+    Value *Idx = ConstantInt::get(T_size, offsetof(jl_datatype_t, abstract));
+
+    Value *abstract = tbaa_decorate(tbaa_const,
+            ctx.builder.CreateLoad(T_int8, ctx.builder.CreateGEP(T_int8, Ptr, Idx)));
```
Comment (tkelman) [inline src/cgutils.cpp]: should be 4 space indent, not 8

File: src/ccall.cpp (hunk 1 above)
Comment (vtjnash) [inline src/ccall.cpp]: builder doesn't exist anymore

File: src/ccall.cpp (hunk 2 above)
Comment (vtjnash) [inline src/ccall.cpp]: I guess I missed one. Not a bad record though, eh?

File: src/cgutils.cpp [hunk 7]
```diff
@@ -973,23 +962,23 @@ static inline Instruction *maybe_mark_load_dereferenceable(Instruction *LI, bool
     return maybe_mark_load_dereferenceable(LI, can_be_null, size);
 }
 
-static void null_pointer_check(Value *v, jl_codectx_t *ctx)
+static void null_pointer_check(jl_codectx_t &ctx, Value *v)
 {
-    raise_exception_unless(builder.CreateICmpNE(v,Constant::getNullValue(v->getType())),
-                           literal_pointer_val(jl_undefref_exception), ctx);
+    raise_exception_unless(ctx,
+            ctx.builder.CreateICmpNE(v, Constant::getNullValue(v->getType())),
```
Comment (tkelman) [inline src/cgutils.cpp]: 4 space indent, not 8

File: src/ccall.cpp (hunk 3 above)
Comment (vtjnash) [inline src/ccall.cpp]: I use this pattern fairly often and try to be consistent about it: this indentation pattern makes it indented enough to be visually distinct and easy to count arguments (each gets its own line), while helping minor refactoring (like changing names or other arguments) to have a minimal diff.

File: src/ccall.cpp (hunk 4 above)
Comment (vtjnash) [inline src/ccall.cpp]: visually distinct from control flow

File: src/cgutils.cpp [hunk 8]
```diff
@@ -1424,37 +1415,37 @@ static jl_cgval_t emit_getfield_knownidx(const jl_cgval_t &strct, unsigned idx,
         bool isboxed;
         Type *lt = julia_type_to_llvm((jl_value_t*)jt, &isboxed);
         if (isboxed) {
-            Value *ptr = decay_derived(data_pointer(strct, ctx, T_pint8));
+            Value *ptr = decay_derived(data_pointer(ctx,strct, T_pint8));
```
Comment (tkelman) [inline src/cgutils.cpp]: keep spaces after commas

File: src/cgutils.cpp [hunk 9]
```diff
@@ -1522,61 +1512,63 @@ static void maybe_alloc_arrayvar(int s, jl_codectx_t *ctx)
...
-    Value *t = boxed(tinfo, ctx);
-    int o = offsetof(jl_array_t, nrows)/sizeof(void*) - 1;
+    Value *t = boxed(ctx, tinfo);
+    int o = offsetof(jl_array_t, nrows) / sizeof(void*) - 1;
     MDNode *tbaa = arraytype_constshape(tinfo.typ) ? tbaa_const : tbaa_arraysize;
-    return emit_nthptr_recast(t, builder.CreateAdd(dim,
-                                                   ConstantInt::get(dim->getType(), o)),
-                              tbaa, T_psize);
+    return emit_nthptr_recast(ctx,
+            t,
```
Comment (tkelman) [inline src/cgutils.cpp]: 4 space indent, not 8

File: src/cgutils.cpp (hunk 5 above)
Comment (vtjnash) [inline src/cgutils.cpp]: simpler to just bitcast once

File: src/cgutils.cpp (hunk 8 above)
Comment (vtjnash) [inline src/cgutils.cpp]: I blame `sed` :P

I attempted to clean everything up as I went, but let me know if I missed any others. Although, really would probably more effective for your time to review #15591 and push that forwards towards merging.

File: src/ccall.cpp (hunk 4 above)
Comment (tkelman) [inline src/ccall.cpp]: I agree with an extra level of indentation for line continuations of block openers - signatures, if conditions, etc. But not for call sites in normal expressions within a block.

File: src/ccall.cpp (hunk 3 above)
Comment (tkelman) [inline src/ccall.cpp]: visually distinct from what? it obviously isn't control flow by virtue of calling something else and not introducing a new block.

Comment (tkelman) [inline src/ccall.cpp]: github and command-line git are both quite good at showing intra-line changes with `--color-words`, this "one line per argument" thing means code takes up a lot more space and takes more scrolling to read than it really needs to

File: src/codegen.cpp [hunk 10]
```diff
@@ -3250,8 +3234,11 @@ static jl_cgval_t emit_call(jl_expr_t *ex, jl_codectx_t *ctx)
         std::map<jl_fptr_t,Function*>::iterator it = builtin_func_map.find(jl_get_builtin_fptr(f));
         if (it != builtin_func_map.end()) {
             theFptr = (*it).second;
-            result = mark_julia_type(emit_jlcall(theFptr,
-                maybe_decay_untracked(V_null), &args[1], nargs, ctx), true, expr_type(expr,ctx), ctx);
+            result = mark_julia_type(
+                    ctx,
```
Comment (tkelman) [inline src/codegen.cpp]: 4 space indent, not 8

File: src/codegen.cpp [hunk 11]
```diff
@@ -4017,77 +4005,79 @@ static jl_cgval_t emit_expr(jl_value_t *expr, jl_codectx_t *ctx)
...
+        Value *typ = boxed(ctx, emit_expr(ctx, args[0]));
+        Value *val = emit_jlcall(ctx, jlnew_func, typ, &args[1], nargs - 1);
+        return mark_julia_type(ctx, val, true, ty);
     }
     else if (head == exc_sym) { // *ptls->exception_in_transit
-        return mark_julia_type(builder.CreateLoad(emit_exc_in_transit(ctx),
-                                                  /*isvolatile*/true),
-                               true, jl_any_type, ctx);
+        return mark_julia_type(ctx,
+                ctx.builder.CreateLoad(emit_exc_in_transit(ctx), /*isvolatile*/true),
```
Comment (tkelman) [inline src/codegen.cpp]: 4 space indent, not 8

File: src/cgutils.cpp (hunk 5 above)
Comment (tkelman) [inline src/cgutils.cpp]: seems like a functional change then

Comment (vtjnash) [inline src/cgutils.cpp]: nah, just less code for functional equivalence

File: src/codegen.cpp (hunk 11 above)
Comment (vtjnash) [inline src/codegen.cpp]: seriously, #15591 would be much more efficient at this. It's not worth your time to point these out or my time to fix them, when the computer can do both instantly.

File: src/intrinsics.cpp [hunk 12]
```diff
@@ -559,187 +561,190 @@ static Value *generic_fpext(Type *to, Value *x, jl_codectx_t *ctx)
...
         assert(!isboxed);
-        thePtr = emit_unbox(ptrty, e, e.typ);
+        thePtr = emit_unbox(ctx, ptrty, e, e.typ);
         if (ety == (jl_value_t*)jl_any_type) {
             // unsafe_store to Ptr{Any} is allowed to implicitly drop GC roots.
-            Instruction *store = builder.CreateAlignedStore(
-              emit_pointer_from_objref(boxed(x, ctx, false)),
-                builder.CreateGEP(thePtr, im1), align_nb);
+            Instruction *store = ctx.builder.CreateAlignedStore(
+              emit_pointer_from_objref(ctx, boxed(ctx, x, false)),
```
Comment (tkelman) [inline src/intrinsics.cpp]: should be 4 space indent

File: src/intrinsics.cpp [hunk 13]
```diff
@@ -559,187 +561,190 @@ static Value *generic_fpext(Type *to, Value *x, jl_codectx_t *ctx)
...
+static Value *emit_checked_srem_int(jl_codectx_t &ctx, Value *x, Value *den)
 {
     Type *t = den->getType();
-    raise_exception_unless(builder.CreateICmpNE(den, ConstantInt::get(t,0)),
-                           literal_pointer_val(jl_diverror_exception), ctx);
-    BasicBlock *m1BB = BasicBlock::Create(jl_LLVMContext,"minus1",ctx->f);
-    BasicBlock *okBB = BasicBlock::Create(jl_LLVMContext,"oksrem",ctx->f);
-    BasicBlock *cont = BasicBlock::Create(jl_LLVMContext,"after_srem",ctx->f);
+    raise_exception_unless(ctx,
+            ctx.builder.CreateICmpNE(den, ConstantInt::get(t, 0)),
```
Comment (tkelman) [inline src/intrinsics.cpp]: should be 4 space indent, not 8

File: src/codegen.cpp (hunk 11 above)
Comment (tkelman) [inline src/codegen.cpp]: This took an hour on an exceptionally large diff. 15591 hasn't gone anywhere in a year. Yes it'd be great if someone wants to follow through with that and finish it. For now code style is enforced through review like it's always been.

Comment (vtjnash) [inline src/codegen.cpp]: > 15591 hasn't gone anywhere in a year

er, no. I reviewed it about 6 months ago and it's now been waiting on a second reviewer who has a spare hour or so to inspect it. Anyways, I've pushed an updated copy https://github.com/JuliaLang/julia/pull/22660

**Comment by vtjnash** [comment]
Indentation to be addressed by #22660
